Make the polyfill work in IE & Edge

Lodin opened the pull request:
Resolves #12.
Resolves #25.
Fixes #29. 

This PR provides support for IE in the polyfill.

@calebdwilliams, sorry for jumping in, I know that you're working on the IE support but you mentioned that you don't have IE11 machine now. So I decided to start working on it in parallel since I have both IE11 and Edge on my machine. Please, feel free to add any solution you prefer.

PR is still in the early stage, a lot of tests are red in IE and Edge now. It may take some time to find a fix all of them.

calebdwilliams commented:
I think this makes sense if you have immediate access to IE11. As for polyfills, IE supports `WeakMap` and `Map` (but not iterable Maps). I would prefer using Rollup to bundle here (or just npm scripts) because the output will be cleaner.

calebdwilliams commented:
I did also want to say this package should not include its own polyfills for Promise or Map to ensure compatibility with other tools that might be used. Instead please add notes inside the README that note what will be needed.

For example, if I'm running an Angular application, I want the Promise used in this polyfill to be the same one loaded by Angular (or any other framework). I can still be convinced on the `Node.isConnected` polyfill, but that has less impact on consuming projects.

Lodin commented:
> As for polyfills, IE supports WeakMap and Map (but not iterable Maps).

Oh, does it? I thought that it is a ES6 thing for some reason. That's the good news.

> I would prefer using Rollup to bundle here (or just npm scripts) because the output will be cleaner.

You mean to transpile the existing code? Well, we can do that but, AFAIK, all polyfills are written with ES5 to make manual tweaks (I did them in the code as well) and get the resulting code size smaller. I personally would prefer to follow this tradition. Would do you think about it? 

> I did also want to say this package should not include its own polyfills for Promise or Map to ensure compatibility with other tools that might be used.

Sure, I completely agree here. Polyfills should not be dependencies, they are the support code that the end user adds if necessary.

> I can still be convinced on the `Node.isConnected`

BTW, I think I found a workaround to not use the polyfill at all. The idea is that if we have `ShadowRoot` available (I mean, the native `ShadowRoot`), the browser should definitely support `isConnected`. And if not, we can just run the simple `document.contains` check.

Lodin commented:
@calebdwilliams, are you aware how to work with the ShadyCSS? I'm kinda stuck with it. It has its own implementation for `adoptedStyleSheet`, so all we need to do if it is defined in `window` is to rely on its API. However, I cannot find any appropriate docs or tutorial on how to use this API. Do you know something?

calebdwilliams commented:
Really good thought on `ShadowRoot` and `Node.isConnected`. I like that approach a lot. 

Not sure what you mean here:
>You mean to transpile the existing code? Well, we can do that but, AFAIK, all polyfills are written with ES5 to make manual tweaks (I did them in the code as well) and get the resulting code size smaller. I personally would prefer to follow this tradition. What do you think about it?

Lodin commented:
@calebdwilliams, hmm, I'm about your mention of the Rollup. I might mistunderstood you; what did you mean regarding Rollup?

calebdwilliams commented:
So thinking about this a bit, I believe it would be better to keep the src file in ES6+ and Babel down to IE11 support (primarily to keep the code as maintainable as possible). My ideal setup would use Rollup and split out the source into src and maybe divide the code a bit. Then use Babel to output the file in the project’s root. I realize you have done a lot of work on this already, let’s pause though and consider what approach is s most sustainable long term.

Lodin commented:
Well, I don't think that keeping the code in ES5 would be harder than work with it in ES6. I rewrote it, and I don't see much difference. The most visible changes are the following:
* `const`/`let` -> `var`. Well, that's a mostly cosmetic change.
* `for..of` -> `forEach`. This change should be done in both variants to support IE11 native Map `forEach`.
* No destructuring. That's the most painful, probably, but I don't see it as a show-stopper. Now it is wordier, but still quite understandable.
* No arrow functions. It brings some troubles with `this` but it is not very popular in the code, so I think I've already fixed every issue.

So, I don't see a lot of disadvantages while there is some advantages:
* No building system, no additional configuration, the file can be used in any browser with any loading system. 
* While we work with ES5, we understand that there could be some caveats like [this one](https://github.com/Lodin/construct-style-sheets/blob/feat/adopt-ie11/adoptedStyleSheets.js#L288). In ES6 it would be surprising because ES6 is supposed to work with polyfills to fix this type of issues (well, at least I think this way).
* We can optimize some ES5 code better than Babel would do it. E.g., [here](https://github.com/Lodin/construct-style-sheets/blob/feat/adopt-ie11/adoptedStyleSheets.js#L92) I just use the `arguments` object when Babel inserts `for` loop to rewrite arguments to an args array. Well, that's not a lot of work nor the bytes, but with the cases above it may play its role.

Plus, as I mentioned above, the polyfills are primarily written in ES5 to be consumed by old browsers directly. I believe we should follow this tradition.

That are my ideas regarding this question. I do not totally disagree with ES6 solution, but while debugging the code in Edge and IE11, I found that it would be way easier to work with ES5 code than a transpiled ES6 (especially it is connected with `async`-`await`). That was the reason to rewrite tests to ES5 as well.

Lodin commented:
Ok, from the #30 it looks like that we would need more files to properly support `CSSStylesSheet` rules. So I agree with ES6 build and Rollup.

Lodin commented:
Hmm, I would say that I'm kinda stuck with the ShadyCSS. I'm not sure it can be properly integrated, all my attempts are like making the polyfill more complicated than it should be.

I'm going to push all changes in the current state, then purge all ShadyCSS-related things and then make PR available. For IE it will work only for `document` but not for polyfilled custom elements. Tests will be passed only in Firefox, Chrome and Safari (or we can disable Custom Element tests for IE & Edge).

So there will be a history of ShadyCSS attempts and if someone would be able to make it work, it would be great.

Lodin commented:
By the way, I'm not sure we should work on it since ShadyCSS has its own implementation already. Support of adopted style sheets in shadow roots is tightly coupled with supporting web components, and it is definitely out of scope for this polyfill.

Lodin commented:
Hmm, the version of Firefox and Chrome in travis is rather old (56 when the current is 68), it might be a reason why tests fail. Because I have all of them passing on my local Firefox 68.

calebdwilliams commented:
Sorry I've been out of the pocket over the weekend, it was a holiday here in the US yesterday, but I'm back online. Can you get some info about _why_ the tests are failing? I'll see if I can't figure out how to upgrade Firefox in Travis (or if that's even possible).

Lodin commented:
@calebdwilliams, probably, it does not support Web Components. AFAIK, Firefox supports them [starting from the version 63](https://hacks.mozilla.org/2018/11/the-power-of-web-components/) (October of 2018) while Travis suggests version 56.

I really don't understand why it does provide such old browsers. My project runs on the [Chrome 76](https://travis-ci.org/corpusculejs/corpuscule/builds/575283217), and my [.travis.yml](https://github.com/corpusculejs/corpuscule/blob/3e45860910cced36dfd5245338bf3819bddb3140/.travis.yml) file does not have much difference from yours. The only big difference I can see is that your project runs on https://travis-ci.com (the new one) and my - on the https://travis-ci.org (the old one). But I don't see any particular reason to load such different versions.

However, let's try setting Firefox from `stable` to `latest` (I've seen it in the [Travis docs](https://docs.travis-ci.com/user/firefox))

Lodin commented:
By the way, what do you think about [my suggestion](https://github.com/calebdwilliams/construct-style-sheets/pull/28#issuecomment-526963542) of not supporting ShadyCSS at all since it has its own support? We would be able to support only the `document` level of `adoptedStyleSheet` and it may play fine with ShadyCSS implementation of Constructible Style Sheets.

calebdwilliams commented:
Let's put ShadyCSS support on the back burner for now. That might be something to revisit in the future, but that is tangential to the primary goal of this PR.

As for Firefox, please feel free to update the Travis config. I saw both versions listed on the Travis documentation, so it could have been a mis-type somewhere.

Lodin commented:
I already [purged](https://github.com/calebdwilliams/construct-style-sheets/pull/28/commits/209fe3bd9ef7e1b9a523a5058cf3b4f256682a9c) all attempts to get ShadyCSS working. 

> Before this update is ready, I'd like these commits to be squashed, too.

Maybe, it would be better to squash the whole PR? You can do it via the Github interface:
<img src="https://user-images.githubusercontent.com/8037479/64190091-44623a00-ce7e-11e9-9817-60f701958e98.png" width="300px">

Lodin commented:
Ok, looks like it works :) Tests are passed

Lodin commented:
Oh wait, one more thing to remove

Lodin commented:
Ok, this PR is ready, let's squash it and merge 🙂

calebdwilliams commented on scripts/build.js:
```diff
@@ -0,0 +1,26 @@
+    console.error(e.stack);
+    process.exit(1);
+  }
+})();
```
What is the purpose of this file?

Lodin replied:
Unfortunately, buy the rollup itself I don't see any possibility to inject the check for implemented `adoptStyleSheets` in browser. I can do it in the index for, but also a lot it code like `WeakMap` creation will be excuted without need. So I used the direct injection of the code into the result of bundling. Now it detects implementation before any code runs.

calebdwilliams replied:
Assuming `index.js` gets bundled as part of the build, the 

```javascript
if ('adoptedStyleSheets' in document) {
  return;
}```

The build step should really just be `[npx]? rollup -c`.

Lodin replied:
The problem is that index is the last file to bundle. So we will have something like that:
```js
(function() {
  'use strict'

  var adoptedSheetsRegistry = new WeakMap();
  var sheetMetadataRegistry = new WeakMap();
  var locationRegistry = new WeakMap();

  // A LOT of code, function delarations, etc.
  // Basically, there will be the whole polyfill code, because Rollup includes file by resolving 
  // their imports, where the entry point always comes at the end

  // Because we cannot `return` at the root of ES module
  if (!('adoptedStyleSheets' in document)) {
    // The small amount of content that is placed in `src/index.js`
    updatePrototype(OldCSSStyleSheet.prototype);

    window.CSSStyleSheet = ConstructStyleSheet;

    initAdoptedStyleSheets();

    if (document.readyState === 'loading') {
      document.addEventListener('DOMContentLoaded', initPolyfill);
    } else {
      initPolyfill();
    }
  }
})();
```
So, e.g., Chrome users will receive a polyfill that does a lot of work even if it is unnecessary. I would like to avoid this kind of situation. However, the only way I found is to inject the code after the build. I understand that it might look ugly, but it works.

Some other ways don't look like a solution as well.
- We can drop rollup and get back to the single file (however, it would harm maintainability).
- We can leave the detection to the user (which is against the common polyfill traditions).
- We can put detection code to every file (which would produce a lot of unnecessary code and if checks).

I chose the easiest and most efficient way. I consider this approach as using a rollup plugin. 

Hmm, probably, it would be better to put the detection code to the JS file in `src` instead of a string, it may improve maintainability by noting that the detection code does not appear from somewhere.

Lodin replied:
Oh, I cannot put it to a file. It causes eslint to yell at me that `return` is used outside of a function. Well, then let it be a string.

calebdwilliams replied:
Got it. I'm not a huge fan of the approach, but can't think of anything better. This is looking really good, however with the caveat that `instanceof` doesn't work in IE11. 

I think that is acceptable for IE11 for the time being if we add that as a disclaimer in the README. I've glanced over the code and have no real issues, I'll start a deep dive now, but if you wouldn't mind adding that language to the README, I'll review and start the merge/release cycle.

Lodin replied:
> however with the caveat that `instanceof` doesn't work in IE11

Yeah, but Babel has the [@babel/plugin-transform-instanceof](https://www.npmjs.com/package/@babel/plugin-transform-instanceof) to handle it. I used it for tests. We can mention that this plugin might be necessary in the docs.

Lodin commented on karma.conf.js:
```diff
@@ -1,6 +1,8 @@
 // Karma configuration
 // Generated on Sun Jan 20 2019 23:06:22 GMT-0600 (CST)
 
+const webpack = require('./webpack.config');
```
Hmm, `karma-webpack` uses webpack dev server to run tests, so no bundle file in the middle. We, probably, can use the result of `npm run build`, but I've never thought this way; it looks a bit strange for me to include different stage into testing. But your idea looks valid, let me change the testing config then.

Lodin replied:
Fixed

Lodin commented on README.md:
````diff
@@ -24,7 +24,17 @@ This polyfill should work as-is in modern browsers, however, to support IE11, yo
+or
+```javascript
+import 'https://unpkg.com/construct-style-sheets-polyfill';
+```
````
Hmm, are you sure? I just followed the usual pattern I've seen in the majority of the project if worked with. All of them places only one or two lines of commands: for npm and sometimes yarn.

calebdwilliams replied:
Yes. Our industry has enough hidden knowledge as it is. Adding one line of instruction for the uninitiated could make a difference for someone. Besides, anyone who knows would only need the package name, we _could_ write something like: `This package is available on npm under the name construct-style-sheet-polyfill and can be installed with npm, yarn, unkpg or however else you consume dependencies.`. Adding a like to npm docs is just a courtesy.

Lodin replied:
Fixed.

calebdwilliams commented:
I was also thinking [`Element.insertAdjacentElement`](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement) with `'afterbegin'` might be a better solution for our `Element.append` approach.

Lodin commented:
Unfortunately, `Element.insertAdjacentElement` works only with objects that implement `Element` interface, not the `Node`. So, we cannot use it with the `DocumentFragment`. I don't know why it designed to work this way, such a disappointment.

calebdwilliams commented on test/_index.html:
```diff
@@ -5,6 +5,6 @@
     <h1>Hello world</h1>
 </body>
-<script src="../adoptedStyleSheets.js"></script>
+<script type="module" src="../src/index.js"></script>
```
Better to load from dist or set up a script to run this from localhost otherwise this doesn't work from `file://`

Lodin replied:
Oh yeah, I forgot about this part.

Lodin replied:
Fixed

calebdwilliams replied:
I mean, it's still `[type=module]` which won't load from `file://`

calebdwilliams commented on test/index.html:
```diff
@@ -3,12 +3,12 @@
         <title>Constructed style sheets test</title>
         <meta charset="UTF-8">
-        <script src="../adoptedStyleSheets.js"></script>
+        <script src="../src/index.js" type="module"></script>
```
Better to load from dist or set up a script to run this from localhost otherwise this doesn't work from `file://`

Lodin replied:
Fixed

calebdwilliams commented on README.md:
```diff
@@ -10,50 +10,89 @@ Currently [Mozilla is considering implementation of the feature](https://github.
+required.
+
+For browsers that do not support the web components specification (currently
+IE 11 and Edge) only the document-level style sheets adoption works. 
```
And checking `instanceof` will not work in IE11 due to `Symbol` not being natively supported.

Lodin replied:
Hmm, I just fixed it and the previous one. It looks the following now:
```html
<script src="../dist/adoptedStyleSheets.js"></script>
```
and 
```markdown
## Support
This polyfill supports all modern browsers and IE 11.

For browsers that do not support the web components specification (currently
IE 11 and Edge) only the document-level style sheets adoption works.

### IE 11

To make this polyfill work with IE 11 you need the following tools:
- `Symbol` polyfill (with support for `Symbol.hasInstance`).
- [@babel/plugin-transform-instanceof](https://www.npmjs.com/package/@babel/plugin-transform-instanceof)
applied to your code that uses `instanceof` against `CSSStyleSheet`.
```
